flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -529,6 +579,22 @@ public String toString() {
     // Protected methods
     // ---------------------------------------------------------------------------------------------------------
 
+    /**
+     * Checks whether a lock is created for this instance on the passed ZooKeeper node.
+     *
+     * @param rootPath The node that shall be checked.
+     * @return {@code true} if the lock exists; {@code false} otherwise.
+     */
+    private boolean hasLock(String rootPath) throws Exception {
+        final String normalizedRootPath = normalizePath(rootPath);
+        try {
+            return client.checkExists().forPath(getLockPath(normalizedRootPath)) != null;
```

Comment:
Under which circumstances can this return null? (as non existent node seems to throw an exception)

Reply:
This was meant for hardening the implementation. In the current setup it cannot happen, because this method is only called by `replace` which is only used in `DefaultJobGraphStore` when the corresponding `JobGraph` is already added to `DefaultJobGraphStore.addedJobGraphs` (which means that the JobGraph was already locked before through `getAndLock` in `DefaultJobGraphStore.recoverJobGraph` or `addAndLock` in `DefaultJobGraphStore.putJobGraph`)

Reply:
This was meant for hardening the implementation. In the current setup it cannot happen, because this method is only called by `replace` which is only used in `DefaultJobGraphStore` when the corresponding `JobGraph` is already added to `DefaultJobGraphStore.addedJobGraphs` (which means that the JobGraph was already locked before through `getAndLock` in `DefaultJobGraphStore.recoverJobGraph` or `addAndLock` in `DefaultJobGraphStore.putJobGraph`)